### ISSUE TITLE
possible numerical issue in UnitGaussianNormalizer

### ISSUE THREAD

**Issue by gokhalen** [neuraloperator/neuraloperator#329]
@zongyi-li @dhpitt 

I'm using the [UnitGaussianNormalizer](https://github.com/neuraloperator/neuraloperator/blob/889ea5115e217d0935e628c7be9517b0e2893e5b/neuralop/datasets/output_encoder.py#L80) class and I think I am running into a numerical issue. 

After I put a breakpoint after line 155 I can see:

```
self.std = 
tensor([[[[[0.3531]]],


         [[[0.1378]]],


         [[[0.2087]]],


         [[[0.2484]]],


         [[[   nan]]]]])
```

However, if I do `torch.std(data_batch,axis=(0,2,3,4),correction=1)` I see

```
tensor([0.3529, 0.1378, 0.2083, 0.2509, 0.0962])
```

Which matches the first 4 entries but not the last. 

```
self.squared_mean - self.mean**2
tensor([[[[[ 0.1247]]],


         [[[ 0.0190]]],


         [[[ 0.0436]]],


         [[[ 0.0617]]],


         [[[-0.0035]]]]])
```

And that last negative number, the result of a catastrophic cancellation somewhere, is the cause of trouble.  As a final note, `torch.mean` matches `self.mean`. Any ideas on how to fix? Should I use `torch.std` instead?

**Comment by JeanKossaifi** [comment]
Hi @gokhalen, thanks for reporting! Could you share a minimal reproducible example? Seems we need better unit-tests for this class. @dhpitt we should look into it, this can have severe impact in applications.

**Comment by gokhalen** [comment]
@JeanKossaifi @dhpitt 

See the small script below, which compares the two methods of computing the stddev. It seems `torch.std` is more reliable than the current method in `neuralop`.  Might be better to compute `mean` and `squared_mean` in double precision and then convert to half precision. 

```
import torch, numpy as np
from neuralop.datasets.output_encoder import UnitGaussianNormalizer
import sys


ncases = 10000
std    = 0.001
mean   = 2.0  # this needs to be about 2 or greater to cause failures
dtype  = 'float32'
torch_dtype = torch.float32
ntrain = 5
nfeat  = 5
nx, ny = 16,16
ntime  = 8
dim    = [0,2,3,4]

neuralop_failures = 0
torch_failures    = 0

for icase in range(ncases):
    # Generate random data and Convert to Torch
    data_np    = np.random.normal(loc=mean,scale=std,size=ntrain*nfeat*nx*ny*ntime).astype(dtype)
    data_np    = data_np.reshape(ntrain,nfeat,nx,ny,ntime)
    data_torch = torch.as_tensor(data_np,dtype=torch_dtype)

    # current method in neuraloperator
    mean_         = torch.mean(data_torch, dim=dim, keepdim=True)
    squared_mean_ = torch.mean(data_torch**2, dim=dim, keepdim=True)
    std_no        = torch.sqrt(squared_mean_ - mean_**2)

    # computing using torch.std
    std_torch     = torch.std(data_torch,dim=dim,correction=0,keepdim=True)

    bool_neuralop_failure = torch.any(torch.isnan(std_no))
    if (bool_neuralop_failure): 
        neuralop_failures +=1
        #print(f'{std_no=}')

    bool_torch_failure = torch.any(torch.isnan(std_torch))
    if (bool_torch_failure):
        torch_failures +=1
        #print(f'{std_torch=}')

    
    print(f'{icase=}')
    if bool_neuralop_failure or bool_torch_failure:
        print(f'{neuralop_failures=}, {torch_failures=}')

print(f'{neuralop_failures=}, {torch_failures=}')
```

**Comment by dhpitt** [comment]
Hi @gokhalen, we updated our method in #331 and added flaky tests to verify correctness.
